streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala

```diff
@@ -0,0 +1,104 @@
+import org.apache.spark.streaming.receiver.Receiver
+
+class InputInfoTrackerSuite extends FunSuite with BeforeAndAfter {
+
+  private var ssc: StreamingContext = _
+
+  before {
+    val conf = new SparkConf().setMaster("local[2]").setAppName("DirectStreamTacker")
+    if (ssc == null) {
+      ssc = new StreamingContext(conf, Duration(1000))
+    }
+  }
+
+  after {
+    if (ssc != null) {
+      ssc.stop()
+      ssc = null
+    }
+  }
+
+  test("test track the number of input stream") {
+    class TestInputDStream extends InputDStream[String](ssc) {
+      def start() { }
+      def stop() { }
+      def compute(validTime: Time): Option[RDD[String]] = ???
+    }
+
+    class TestReceiverInputDStream extends ReceiverInputDStream[String](ssc) {
+      def getReceiver: Receiver[String] = ???
+    }
+
+    // Register input streams
+    val receiverInputStreams = Array(new TestReceiverInputDStream, new TestReceiverInputDStream)
+    val inputStreams = Array(new TestInputDStream, new TestInputDStream, new TestInputDStream)
+
+    assert(ssc.graph.getInputStreams().length == 5)
+    assert(ssc.graph.getReceiverInputStreams().length == 2)
+    assert(ssc.graph.getReceiverInputStreams() === receiverInputStreams)
+    assert(ssc.graph.getInputStreams().map(_.id) === Array(0, 1, 2, 3, 4))
+    assert(receiverInputStreams.map(_.id) === Array(0, 1))
+  }
+
+  test("test report and get InputInfo from InputInfoTracker") {
+    val inputInfoTracker = new InputInfoTracker(ssc)
+
+    val streamId1 = 0
+    val streamId2 = 1
+    val time = Time(0L)
+    val inputInfo1 = InputInfo(time, streamId1, 100L)
+    val inputInfo2 = InputInfo(time, streamId2, 300L)
+    inputInfoTracker.reportInfo(inputInfo1.batchTime, inputInfo1)
+    inputInfoTracker.reportInfo(inputInfo2.batchTime, inputInfo2)
+
+    val batchTimeToInputInfos = inputInfoTracker.getInfo(time)
+    assert(batchTimeToInputInfos.size == 2)
+    assert(batchTimeToInputInfos.keys === Set(streamId1, streamId2))
+    assert(batchTimeToInputInfos(streamId1) === inputInfo1)
+    assert(batchTimeToInputInfos(streamId2) === inputInfo2)
+    assert(inputInfoTracker.getInfo(time)(streamId1) === inputInfo1)
+  }
+
+  test("test cleanup InputInfo from InputInfoTracker") {
+    val inputInfoTracker = new InputInfoTracker(ssc)
+
+    val streamId1 = 0
+    val inputInfo1 = InputInfo(Time(0), streamId1, 100L)
+    val inputInfo2 = InputInfo(Time(1), streamId1, 300L)
+    inputInfoTracker.reportInfo(inputInfo1.batchTime, inputInfo1)
+    inputInfoTracker.reportInfo(inputInfo2.batchTime, inputInfo2)
+
+    inputInfoTracker.cleanup(Time(0))
+    assert(inputInfoTracker.getInfo(Time(0))(streamId1) === inputInfo1)
+    assert(inputInfoTracker.getInfo(Time(1))(streamId1) === inputInfo2)
+
+    inputInfoTracker.cleanup(Time(1))
+    assert(inputInfoTracker.getInfo(Time(0)).get(streamId1) === None)
+    assert(inputInfoTracker.getInfo(Time(1))(streamId1) === inputInfo2)
+  }
```

Comment:
I think we need an end-to-end test as well, to test the code path interacting between JobGenerator, StreamingListener and InputInfoTracker. I thought about what is a simple way to do and here is an idea. Checkpoint StreamingListenerSuite. The first test "batch info" can be extended to test the numRecords. For that to work, the input stream used in these tests is [TestInputDStream](https://github.com/apache/spark/blob/master/streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala#L43). You could extend add a line in the TestSuiteBased.compute to post InputInfo. And then test the info in StreamingListenerSuite. Does that make sense?